server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -376,32 +484,57 @@ remoteField
   -> Maybe G.Description
   -> G.ArgumentsDefinition
   -> G.TypeDefinition [G.Name]
-  -> m (FieldParser n ()) -- TODO return something useful, maybe?
+  -> m (FieldParser n (Field NoFragments G.Name))
 remoteField sdoc fieldName description argsDefn typeDefn = do
   -- TODO add directives
   argsParser <- argumentsParser argsDefn sdoc
   case typeDefn of
     G.TypeDefinitionObject objTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaObject sdoc objTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
-    G.TypeDefinitionScalar (G.ScalarTypeDefinition desc name' _) ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldScalarParser name' desc
+      remoteSchemaObjFields <- remoteSchemaObject sdoc objTypeDefn
+      -- converting [Field NoFragments Name] to (SelectionSet NoFragments G.Name)
+      let remoteSchemaObjSelSet = fmap G.SelectionField <$> remoteSchemaObjFields
+      pure remoteSchemaObjSelSet
+        <&> mkFieldParserWithSelectionSet fieldName description argsParser
+    G.TypeDefinitionScalar scalarTypeDefn ->
+      pure $ mkFieldParserWithoutSelectionSet fieldName description argsParser
+             $ remoteFieldScalarParser scalarTypeDefn
     G.TypeDefinitionEnum enumTypeDefn ->
-      pure $ P.selection fieldName description argsParser $ remoteFieldEnumParser enumTypeDefn
-    G.TypeDefinitionInterface ifaceTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaInterface sdoc ifaceTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
-    G.TypeDefinitionUnion unionTypeDefn -> do
-      remoteSchemaObj <- remoteSchemaUnion sdoc unionTypeDefn
-      pure $ void $ P.subselection fieldName description argsParser remoteSchemaObj
+      pure $ mkFieldParserWithoutSelectionSet fieldName description argsParser
+             $ remoteFieldEnumParser enumTypeDefn
+    G.TypeDefinitionInterface ifaceTypeDefn ->
+      remoteSchemaInterface sdoc ifaceTypeDefn <&>
+        mkFieldParserWithSelectionSet fieldName description argsParser
+    G.TypeDefinitionUnion unionTypeDefn ->
+      remoteSchemaUnion sdoc unionTypeDefn <&>
+        mkFieldParserWithSelectionSet fieldName description argsParser
     _ -> throw400 RemoteSchemaError "expected output type, but got input type"
+  where
+    mkFieldParserWithoutSelectionSet
+      :: G.Name
+      -> Maybe Description
+      -> InputFieldsParser n ()
+      -> Parser 'Both n ()
+      -> FieldParser n (Field NoFragments G.Name)
+    mkFieldParserWithoutSelectionSet fldName desc argsParser outputParser =
```

Comment:
Since `fldName` and `desc` are just `fieldName` and `description`, which are in scope, we can use `fieldName` and `description` instead, instead of passing them as arguments.

Reply:
Yes, I'll make the change